p2p/src/utils/rate_limiter/mod.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(clippy::float_arithmetic)]
+
+use std::time::Duration;
+
+/// Token bucket based rate limiter
+///
+/// See https://en.wikipedia.org/wiki/Token_bucket
+#[derive(Debug)]
+pub struct RateLimiter {
+    rate: f64,
+    tokens: f64,
```

Comment:
Why are tokens a floating point number? It's understandable that the rate is a floating point number. But the tokens? Can't we just scale up to make the minimum unit == 1?

Reply:
I'm not sure how it can be done. The problem is that both `seconds` and `rate` here are floats:
```
self.tokens = f64::min(self.tokens + self.rate * seconds, self.bucket.into());
```
And so `tokens` must be float to be accurate.
It's implemented same way in Bitcoin Core: https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L360

Reply:
I don't know. It sounds appropriate to be an integer. But it's OK.

Just please double-check NaN cases and similar are safe before merging this.

Reply:
It should be OK because there is only multiplication and all values can be trusted.